Unable to build gettext-full undefined reference to `u8_uctomb_aux’

### Describe the bug

make[4]: Entering directory '/home/openwrt/master/openwrt/build_dir/host/libtool-2.4.7'
make[4]: *** [GNUmakefile:108: abort-due-to-no-makefile] Error 1
make[3]: [Makefile:67: /home/openwrt/master/openwrt/build_dir/host/libtool-2.4.7/.preparedc38439611ea6ca76cb81219f56e48ddf_18f1e190c5d53547fed41a3eaa76e9e9] Error 2 (ignored)

.....

libtool: link: x86_64-openwrt-linux-musl-gcc -DINSTALLDIR="/usr/bin" -Os -pipe -fno-caller-saves -fno-plt -fhonour-copts -fmacro-prefix-map=/home/openwrt/master/openwrt/build_dir/target-x86_64_musl/gettext-0.21.1=gettext-0.21.1 -Wformat -Werror=format-security -fstack-protector -D_FORTIFY_SOURCE=1 -Wl,-z -Wl,now -Wl,-z -Wl,relro -DPIC -fpic -znow -zrelro -o .libs/gettext gettext-gettext.o -L/home/openwrt/master/openwrt/staging_dir/toolchain-x86_64_gcc-13.1.0_musl/usr/lib -L/home/openwrt/master/openwrt/staging_dir/toolchain-x86_64_gcc-13.1.0_musl/lib ../gnulib-lib/libgrt.a ../intl/.libs/libintl.so
/home/openwrt/master/openwrt/staging_dir/toolchain-x86_64_gcc-13.1.0_musl/lib/gcc/x86_64-openwrt-linux-musl/13.1.0/../../../../x86_64-openwrt-linux-musl/bin/ld: ../gnulib-lib/libgrt.a(libgrt_a-iconv.o): in function rpl_iconv': iconv.c:(.text+0x34b): undefined reference to u8_mbtoucr'
/home/openwrt/master/openwrt/staging_dir/toolchain-x86_64_gcc-13.1.0_musl/lib/gcc/x86_64-openwrt-linux-musl/13.1.0/../../../../x86_64-openwrt-linux-musl/bin/ld: iconv.c:(.text+0x4ba): undefined reference to `u8_uctomb_aux'
collect2: error: ld returned 1 exit status
make[8]: *** [Makefile:1645: gettext] Error 1
make[8]: Leaving directory '/home/openwrt/master/openwrt/build_dir/target-x86_64_musl/gettext-0.21.1/gettext-runtime/src'
make[7]: *** [Makefile:1643: all-recursive] Error 1
make[7]: Leaving directory '/home/openwrt/master/openwrt/build_dir/target-x86_64_musl/gettext-0.21.1/gettext-runtime'
make[6]: *** [Makefile:1548: all] Error 2
make[6]: Leaving directory '/home/openwrt/master/openwrt/build_dir/target-x86_64_musl/gettext-0.21.1/gettext-runtime'
make[5]: *** [Makefile:406: all-recursive] Error 1
make[5]: Leaving directory '/home/openwrt/master/openwrt/build_dir/target-x86_64_musl/gettext-0.21.1'
make[4]: *** [Makefile:362: all] Error 2
make[4]: Leaving directory '/home/openwrt/master/openwrt/build_dir/target-x86_64_musl/gettext-0.21.1'
make[3]: *** [Makefile:189: /home/openwrt/master/openwrt/build_dir/target-x86_64_musl/gettext-0.21.1/.built] Error 2
make[3]: Leaving directory '/home/openwrt/master/openwrt/package/libs/gettext-full'
time: package/libs/gettext-full/compile#381.42#118.66#437.73
ERROR: package/libs/gettext-full failed to build.
make[2]: *** [package/Makefile:116: package/libs/gettext-full/compile] Error 1
make[2]: *** Waiting for unfinished jobs....
data written to "hd.ids"
log written to "hd.log"

### OpenWrt version

master-218ce40cd7

### OpenWrt target/subtarget

x86_64

### Device

x86_64

### Steps to reproduce

1, ubuntu 22.04
2 time make V=s VERBOSE=1 -j$(nproc) JOBS=$(nproc) 2>&1 | tee build.log | grep -i -E "^make.*(error|[1-9]...Entering dir)"
3, branch master

### Terms

- [X] I am reporting an issue for OpenWrt, not an unsupported fork.

I am experiencing the same issue compiling a full build of master on an aarch64 machine.
A multi-core `make -j 4 world` was first attempted, then after that failed on gettext-full, the package was cleaned and started again singly with `make V=sc package/gettext-full/clean, compile`:
```
Making all in src
make[7]: Entering directory '/mnt/nbd/devel/openwrt/build_dir/target-aarch64_cortex-a53_musl/gettext-0.21.1/gettext-runtime/src'
aarch64-openwrt-linux-musl-gcc -DLOCALEDIR=\"/usr/share/locale\" -DHAVE_CONFIG_H -I. -I..  -I. -I. -I.. -I../intl -I./../intl -I../gnulib-lib -I./../gnulib-lib -I/mnt/nbd/devel/openwrt/staging_dir/toolchain-aarch64_cortex-a53_gcc-12.2.0_musl/usr/include -I/mnt/nbd/devel/openwrt/staging_dir/toolchain-aarch64_cortex-a53_gcc-12.2.0_musl/include/fortify -I/mnt/nbd/devel/openwrt/staging_dir/toolchain-aarch64_cortex-a53_gcc-12.2.0_musl/include  -DINSTALLDIR=\"/usr/bin\" -Os -pipe -mcpu=cortex-a53 -fno-caller-saves -fno-plt -fhonour-copts -ffile-prefix-map=/mnt/nbd/devel/openwrt/build_dir/target-aarch64_cortex-a53_musl/gettext-0.21.1=gettext-0.21.1 -Wformat -Werror=format-security -fstack-protector -D_FORTIFY_SOURCE=1 -Wl,-z,now -Wl,-z,relro -DPIC -fPIC  -c -o gettext-gettext.o `test -f 'gettext.c' || echo './'`gettext.c
/bin/sh ../libtool  --tag=CC   --mode=link aarch64-openwrt-linux-musl-gcc -DINSTALLDIR=\"/usr/bin\" -Os -pipe -mcpu=cortex-a53 -fno-caller-saves -fno-plt -fhonour-copts -ffile-prefix-map=/mnt/nbd/devel/openwrt/build_dir/target-aarch64_cortex-a53_musl/gettext-0.21.1=gettext-0.21.1 -Wformat -Werror=format-security -fstack-protector -D_FORTIFY_SOURCE=1 -Wl,-z,now -Wl,-z,relro -DPIC -fPIC   -L/mnt/nbd/devel/openwrt/staging_dir/toolchain-aarch64_cortex-a53_gcc-12.2.0_musl/usr/lib -L/mnt/nbd/devel/openwrt/staging_dir/toolchain-aarch64_cortex-a53_gcc-12.2.0_musl/lib -znow -zrelro  -o gettext gettext-gettext.o ../gnulib-lib/libgrt.a ../intl/libintl.la     
libtool: link: aarch64-openwrt-linux-musl-gcc "-DINSTALLDIR=\"/usr/bin\"" -Os -pipe -mcpu=cortex-a53 -fno-caller-saves -fno-plt -fhonour-copts -ffile-prefix-map=/mnt/nbd/devel/openwrt/build_dir/target-aarch64_cortex-a53_musl/gettext-0.21.1=gettext-0.21.1 -Wformat -Werror=format-security -fstack-protector -D_FORTIFY_SOURCE=1 -Wl,-z -Wl,now -Wl,-z -Wl,relro -DPIC -fPIC -znow -zrelro -o .libs/gettext gettext-gettext.o  -L/mnt/nbd/devel/openwrt/staging_dir/toolchain-aarch64_cortex-a53_gcc-12.2.0_musl/usr/lib -L/mnt/nbd/devel/openwrt/staging_dir/toolchain-aarch64_cortex-a53_gcc-12.2.0_musl/lib ../gnulib-lib/libgrt.a ../intl/.libs/libintl.so
/mnt/nbd/devel/openwrt/staging_dir/toolchain-aarch64_cortex-a53_gcc-12.2.0_musl/lib/gcc/aarch64-openwrt-linux-musl/12.2.0/../../../../aarch64-openwrt-linux-musl/bin/ld: ../gnulib-lib/libgrt.a(libgrt_a-iconv.o): in function `rpl_iconv':
iconv.c:(.text+0x3a4): undefined reference to `u8_mbtoucr'
/mnt/nbd/devel/openwrt/staging_dir/toolchain-aarch64_cortex-a53_gcc-12.2.0_musl/lib/gcc/aarch64-openwrt-linux-musl/12.2.0/../../../../aarch64-openwrt-linux-musl/bin/ld: iconv.c:(.text+0x3a8): undefined reference to `u8_mbtoucr'
/mnt/nbd/devel/openwrt/staging_dir/toolchain-aarch64_cortex-a53_gcc-12.2.0_musl/lib/gcc/aarch64-openwrt-linux-musl/12.2.0/../../../../aarch64-openwrt-linux-musl/bin/ld: iconv.c:(.text+0x3e4): undefined reference to `u8_uctomb_aux'
/mnt/nbd/devel/openwrt/staging_dir/toolchain-aarch64_cortex-a53_gcc-12.2.0_musl/lib/gcc/aarch64-openwrt-linux-musl/12.2.0/../../../../aarch64-openwrt-linux-musl/bin/ld: iconv.c:(.text+0x3e8): undefined reference to `u8_uctomb_aux'
collect2: error: ld returned 1 exit status
make[7]: *** [Makefile:1642: gettext] Error 1
```

--- a/package/libs/gettext-full/Makefile
+++ b/package/libs/gettext-full/Makefile
@@ -22,10 +22,10 @@ PKG_LICENSE:=GPL-3.0-or-later
 PKG_CPE_ID:=cpe:/a:gnu:gettext
 
 PKG_INSTALL:=1
-PKG_BUILD_DEPENDS:=gettext-full/host libxml2
+PKG_BUILD_DEPENDS:=gettext-full/host libunistring libxml2
 PKG_BUILD_PARALLEL:=0
 
-HOST_BUILD_DEPENDS:=gperf/host libxml2/host
+HOST_BUILD_DEPENDS:=gperf/host libunistring/host libxml2/host
 HOST_BUILD_PARALLEL:=0
 
 PKG_SUBDIRS:= \
@@ -87,6 +87,7 @@ CONFIGURE_ARGS += \
 	--with-included-gettext \
 	--without-libintl-prefix \
 	--without-libexpat-prefix \
+	--with-libunistring-prefix=$(STAGING_DIR)/usr \
 	--with-libxml2-prefix=$(STAGING_DIR) \
 	--without-emacs
 
@@ -98,6 +99,7 @@ HOST_CONFIGURE_ARGS += \
 	--disable-java \
 	--disable-openmp \
 	--without-emacs \
+	--with-libunistring-prefix=$(STAGING_DIR_HOSTPKG) \
 	--with-libxml2-prefix=$(STAGING_DIR_HOSTPKG)
 
 HOST_CONFIGURE_VARS += \

--- a/package/libs/gettext-full/patches/200-libunistring-missing-link.patch
+++ b/package/libs/gettext-full/patches/200-libunistring-missing-link.patch
@@ -1,0 +1,21 @@
+--- a/autogen.sh
++++ b/autogen.sh
+@@ -78,6 +78,7 @@ if ! $skip_gnulib; then
+     getopt-gnu
+     gettext-h
+     havelib
++    libunistring-optional
+     memmove
+     noreturn
+     progname
+--- a/gettext-runtime/src/Makefile.am
++++ b/gettext-runtime/src/Makefile.am
+@@ -40,7 +40,7 @@ envsubst_SOURCES = envsubst.c
+ 
+ # Link dependencies.
+ # Need @LTLIBICONV@ because striconv.c uses iconv().
+-LDADD = ../gnulib-lib/libgrt.a @LTLIBINTL@ @LTLIBICONV@ $(WOE32_LDADD)
++LDADD = ../gnulib-lib/libgrt.a $(LTLIBUNISTRING) @LTLIBINTL@ @LTLIBICONV@ $(WOE32_LDADD)
+ 
+ # Specify installation directory, for --enable-relocatable.
+ gettext_CFLAGS = -DINSTALLDIR=\"$(bindir)\"

--- a/package/libs/libunistring/Makefile
+++ b/package/libs/libunistring/Makefile
@@ -1,0 +1,64 @@
+include $(TOPDIR)/rules.mk
+
+PKG_NAME:=libunistring
+PKG_VERSION:=1.1
+PKG_RELEASE:=1
+
+PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.xz
+PKG_SOURCE_URL:=@GNU/$(PKG_NAME)
+PKG_HASH:=827c1eb9cb6e7c738b171745dac0888aa58c5924df2e59239318383de0729b98
+
+PKG_BUILD_PARALLEL:=1
+PKG_INSTALL:=1
+
+PKG_LICENSE:=GPL-3.0
+PKG_LICENSE_FILES:=COPYING
+PKG_CPE_ID:=cpe:/a:gnu:$(PKG_NAME)
+
+include $(INCLUDE_DIR)/host-build.mk
+include $(INCLUDE_DIR)/package.mk
+
+define Package/libunistring
+  SECTION:=libs
+  CATEGORY:=Libraries
+  TITLE:=libunistring
+  URL:=https://www.gnu.org/software/libunistring
+endef
+
+define Package/libunistring/description
+  This library provides functions for manipulating Unicode strings
+  and for manipulating C strings according to the Unicode standard.
+endef
+
+HOST_CFLAGS += $(HOST_FPIC)
+
+TARGET_CFLAGS += $(FPIC)
+
+HOST_CONFIGURE_ARGS += \
+	--enable-static \
+	--disable-shared \
+	--without-libiconv-prefix \
+	--with-pic
+
+CONFIGURE_ARGS += \
+	--enable-static \
+	--enable-shared \
+	--without-libiconv-prefix \
+	--with-pic
+
+define Build/InstallDev
+	$(INSTALL_DIR) $(1)/usr/include
+	$(CP) $(PKG_INSTALL_DIR)/usr/include/*.h $(1)/usr/include/
+	$(INSTALL_DIR) $(1)/usr/include/unistring
+	$(CP) $(PKG_INSTALL_DIR)/usr/include/unistring/*.h $(1)/usr/include/unistring/
+	$(INSTALL_DIR) $(1)/usr/lib
+	$(CP) $(PKG_INSTALL_DIR)/usr/lib/libunistring.{a,so*} $(1)/usr/lib/
+endef
+
+define Package/libunistring/install
+	$(INSTALL_DIR) $(1)/usr/lib
+	$(CP) $(PKG_INSTALL_DIR)/usr/lib/libunistring.so.* $(1)/usr/lib/
+endef
+
+$(eval $(call HostBuild))
+$(eval $(call BuildPackage,libunistring))

